main.cpp

```diff
@@ -1,109 +1,228 @@
+
+        /* opening the socket only allocates resources */
+        result = socket.open(_net);
+        if (result != 0) {
+            printf("Error! socket.open() returned: %d\r\n", result);
+            return;
+        }
 
-    result = net->connect();
-    if (result != 0) {
-        printf("Error! net->connect() returned: %d\n", result);
-        return result;
+        /* now we have to find where to connect */
+        SocketAddress address;
+
+        if (!resolve_hostname(address)) {
+            return;
+        }
+
+        /* set standard HTTP port */
+        address.set_port(80);
+
+        /* finally connect */
+        result = socket.connect(address);
+        if (result != 0) {
+            printf("Error! socket.connect() returned: %d\r\n", result);
+            return;
+        }
+
+        /* exchange an HTTP request and response */
+
+        if (!send_http_request()) {
+            return;
+        }
+
+        if (!receive_http_response()) {
+            return;
+        }
+
+        printf("Demo concluded successfully \r\n");
     }
 
-    // Show the network address
-    SocketAddress a;
-    net->get_ip_address(&a);
-    printf("IP address: %s\n", a.get_ip_address() ? a.get_ip_address() : "None");
-    net->get_netmask(&a);
-    printf("Netmask: %s\n", a.get_ip_address() ? a.get_ip_address() : "None");
-    net->get_gateway(&a);
-    printf("Gateway: %s\n", a.get_ip_address() ? a.get_ip_address() : "None");
+private:
+    bool resolve_hostname(SocketAddress &address)
+    {
+        const char hostname[] = MBED_CONF_APP_HOSTNAME;
+
+        /* get the host address */
+        printf("\nResolve hostname %s\r\n", hostname);
+        nsapi_size_or_error_t result = _net->gethostbyname(hostname, &address);
+        if (result != 0) {
+            printf("Error! gethostbyname(%s) returned: %d\r\n", hostname, result);
+            return false;
+        }
 
-    // Open a socket on the network interface, and create a TCP connection to ifconfig.io
-    TCPSocket socket;
-    // Send a simple http request
-    char hostname[] = "ifconfig.io";
-    char sbuffer[] = "GET / HTTP/1.1\r\nHost: ifconfig.io\r\nConnection: close\r\n\r\n";
-    nsapi_size_t size = strlen(sbuffer);
-
-    result = socket.open(net);
-    if (result != 0) {
-        printf("Error! socket.open() returned: %d\n", result);
-        goto DISCONNECT;
+        printf("%s address is %s\r\n", hostname, (address.get_ip_address() ? address.get_ip_address() : "None") );
+
+        return true;
     }
 
-    // Get the host address
-    printf("\nResolve hostname %s\n", hostname);
-    result = net->gethostbyname(hostname, &a);
-    if (result != 0) {
-        printf("Error! gethostbyname(%s) returned: %d\n", hostname, result);
-        net->disconnect();
-        exit(-1);
+    bool send_http_request()
+    {
+        /* loop until whole request sent */
+        const char buffer[] = "GET / HTTP/1.1\nHost: ifconfig.io\nConnection: close\n\n";
```

Comment:
Why not use a raw string literal to avoid the `\n` ? 
See: https://en.cppreference.com/w/cpp/language/string_literal

Reply:
because I didn't know of that feature but I don't see the point of avoiding \n

Reply:
```
GET / HTTP/1.1
Host: ifconfig.io
Connection: close
```

is more readable, no ? `\n` should be fine for small requests.

Reply:
I don't think so but I don't mind either. Changed to raw string